File: robocode-tankroyale-server/src/main/java/dev/robocode/tankroyale/server/mappers/EventsToEventsMapper.java
package dev.robocode.tankroyale.server.mappers;

import net.robocode2.schema.*;
import net.robocode2.schema.Message.Type;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public final class EventsToEventsMapper {

	private EventsToEventsMapper() {}

	public static List<Event> map(Set<dev.robocode.tankroyale.server.events.Event> events) {
		List<net.robocode2.schema.Event> mappedEvents = new ArrayList<>();
		for (dev.robocode.tankroyale.server.events.Event event : events) {
			mappedEvents.add(map(event));
		}
		return mappedEvents;
	}

	private static Event map(dev.robocode.tankroyale.server.events.Event event) {

		if (event instanceof dev.robocode.tankroyale.server.events.BotDeathEvent) {
			return map((dev.robocode.tankroyale.server.events.BotDeathEvent) event);
		}
		if (event instanceof dev.robocode.tankroyale.server.events.BotHitBotEvent) {
			return map((dev.robocode.tankroyale.server.events.BotHitBotEvent) event);
		}
		if (event instanceof dev.robocode.tankroyale.server.events.BotHitWallEvent) {
			return map((dev.robocode.tankroyale.server.events.BotHitWallEvent) event);
		}
		if (event instanceof dev.robocode.tankroyale.server.events.BulletFiredEvent) {
			return map((dev.robocode.tankroyale.server.events.BulletFiredEvent) event);
		}
		if (event instanceof dev.robocode.tankroyale.server.events.BulletHitBotEvent) {
			return map((dev.robocode.tankroyale.server.events.BulletHitBotEvent) event);
		}
		if (event instanceof dev.robocode.tankroyale.server.events.BulletHitBulletEvent) {
			return map((dev.robocode.tankroyale.server.events.BulletHitBulletEvent) event);
		}
		if (event instanceof dev.robocode.tankroyale.server.events.BulletHitWallEvent) {
			return map((dev.robocode.tankroyale.server.events.BulletHitWallEvent) event);
		}
		if (event instanceof dev.robocode.tankroyale.server.events.ScannedBotEvent) {
			return map((dev.robocode.tankroyale.server.events.ScannedBotEvent) event);
		}
		if (event instanceof dev.robocode.tankroyale.server.events.SkippedTurnEvent) {
			return map((dev.robocode.tankroyale.server.events.SkippedTurnEvent) event);
		}
		throw new IllegalStateException("Event type not handled: " + event.getClass().getCanonicalName());
	}

	private static BotDeathEvent map(dev.robocode.tankroyale.server.events.BotDeathEvent botDeathEvent) {
		BotDeathEvent event = new BotDeathEvent();
		event.setType(Type.BOT_DEATH_EVENT);
		event.setTurnNumber(botDeathEvent.getTurnNumber());
		event.setVictimId(botDeathEvent.getVictimId());
		return event;
	}

	private static BotHitBotEvent map(dev.robocode.tankroyale.server.events.BotHitBotEvent botHitBotEvent) {
		BotHitBotEvent event = new BotHitBotEvent();
		event.setType(Type.BOT_HIT_BOT_EVENT);
		event.setTurnNumber(botHitBotEvent.getTurnNumber());
		event.setBotId(botHitBotEvent.getBotId());
		event.setVictimId(botHitBotEvent.getVictimId());
		event.setEnergy(botHitBotEvent.getEnergy());
		event.setX(botHitBotEvent.getX());
		event.setY(botHitBotEvent.getY());
		event.setRammed(botHitBotEvent.isRammed());
		return event;
	}

	private static BotHitWallEvent map(dev.robocode.tankroyale.server.events.BotHitWallEvent botHitWallEvent) {
		BotHitWallEvent event = new BotHitWallEvent();
		event.setType(Type.BOT_HIT_WALL_EVENT);
		event.setTurnNumber(botHitWallEvent.getTurnNumber());
		event.setVictimId(botHitWallEvent.getVictimId());
		return event;
	}

	private static BulletFiredEvent map(dev.robocode.tankroyale.server.events.BulletFiredEvent bulletFiredEvent) {
		BulletFiredEvent event = new BulletFiredEvent();
		event.setType(Type.BULLET_FIRED_EVENT);
		event.setTurnNumber(bulletFiredEvent.getTurnNumber());
		event.setBullet(BulletToBulletStateMapper.map(bulletFiredEvent.getBullet()));
		return event;
	}

	private static BulletHitBotEvent map(dev.robocode.tankroyale.server.events.BulletHitBotEvent bulletHitBotEvent) {
		BulletHitBotEvent event = new BulletHitBotEvent();
		event.setType(Type.BULLET_HIT_BOT_EVENT);
		event.setTurnNumber(bulletHitBotEvent.getTurnNumber());
		event.setBullet(BulletToBulletStateMapper.map(bulletHitBotEvent.getBullet()));
		event.setVictimId(bulletHitBotEvent.getVictimId());
		event.setDamage(bulletHitBotEvent.getDamage());
		event.setEnergy(bulletHitBotEvent.getEnergy());
		return event;
	}

	private static BulletHitBulletEvent map(dev.robocode.tankroyale.server.events.BulletHitBulletEvent bulletHitBulletEvent) {
		BulletHitBulletEvent event = new BulletHitBulletEvent();
		event.setType(Type.BULLET_HIT_BULLET_EVENT);
		event.setTurnNumber(bulletHitBulletEvent.getTurnNumber());
		event.setBullet(BulletToBulletStateMapper.map(bulletHitBulletEvent.getBullet()));
		event.setHitBullet(BulletToBulletStateMapper.map(bulletHitBulletEvent.getHitBullet()));
		return event;
	}

	private static BulletHitWallEvent map(dev.robocode.tankroyale.server.events.BulletHitWallEvent bulletHitWallEvent) {
		BulletHitWallEvent event = new BulletHitWallEvent();
		event.setType(Type.BULLET_HIT_WALL_EVENT);
		event.setTurnNumber(bulletHitWallEvent.getTurnNumber());
		event.setBullet(BulletToBulletStateMapper.map(bulletHitWallEvent.getBullet()));
		return event;
	}

	private static ScannedBotEvent map(dev.robocode.tankroyale.server.events.ScannedBotEvent scannedBotEvent) {
		ScannedBotEvent event = new ScannedBotEvent();
		event.setType(Type.SCANNED_BOT_EVENT);
		event.setTurnNumber(scannedBotEvent.getTurnNumber());
		event.setScannedByBotId(scannedBotEvent.getScannedByBotId());
		event.setScannedBotId(scannedBotEvent.getScannedBotId());
		event.setEnergy(scannedBotEvent.getEnergy());
		event.setX(scannedBotEvent.getX());
		event.setY(scannedBotEvent.getY());
		event.setDirection(scannedBotEvent.getDirection());
		event.setSpeed(scannedBotEvent.getSpeed());
		return event;
	}

	private static SkippedTurnEvent map(dev.robocode.tankroyale.server.events.SkippedTurnEvent skippedTurnEvent) {
		SkippedTurnEvent event = new SkippedTurnEvent();
		event.setType(Type.SKIPPED_TURN_EVENT);
		event.setTurnNumber(skippedTurnEvent.getTurnNumber());
		return event;
	}
}

File: robocode-tankroyale-server-kotlin/src/main/kotlin/dev/robocode/tankroyale/server/mapper/EventsToEventsMapper.kt
package dev.robocode.tankroyale.server.mapper

import dev.robocode.tankroyale.schema.*
import java.util.ArrayList

object EventsToEventsMapper {
    fun map(events: Set<dev.robocode.tankroyale.server.event.Event>): List<Event> {
        val mappedEvents: MutableList<Event> = ArrayList()
        for (event in events) {
            mappedEvents.add(map(event))
        }
        return mappedEvents
    }

    private fun map(event: dev.robocode.tankroyale.server.event.Event): Event {
        if (event is dev.robocode.tankroyale.server.event.BotDeathEvent) return map(event)
        if (event is dev.robocode.tankroyale.server.event.BotHitBotEvent) return map(event)
        if (event is dev.robocode.tankroyale.server.event.BotHitWallEvent) return map(event)
        if (event is dev.robocode.tankroyale.server.event.BulletFiredEvent) return map(event)
        if (event is dev.robocode.tankroyale.server.event.BulletHitBotEvent) return map(event)
        if (event is dev.robocode.tankroyale.server.event.BulletHitBulletEvent) return map(event)
        if (event is dev.robocode.tankroyale.server.event.BulletHitWallEvent) return map(event)
        if (event is dev.robocode.tankroyale.server.event.ScannedBotEvent) return map(event)
        if (event is dev.robocode.tankroyale.server.event.SkippedTurnEvent) return map(event)
        throw IllegalStateException("Event type not handled: ${event.javaClass.canonicalName}")
    }

    private fun map(botDeathEvent: dev.robocode.tankroyale.server.event.BotDeathEvent): BotDeathEvent {
        val event = BotDeathEvent()
        event.`$type` = Message.`$type`.BOT_DEATH_EVENT
        botDeathEvent.apply {
            event.turnNumber = turnNumber
            event.victimId = victimId.value
        }
        return event
    }

    private fun map(botHitBotEvent: dev.robocode.tankroyale.server.event.BotHitBotEvent): BotHitBotEvent {
        val event = BotHitBotEvent()
        event.`$type` = Message.`$type`.BOT_HIT_BOT_EVENT
        botHitBotEvent.apply {
            event.turnNumber = turnNumber
            event.botId = botId.value
            event.victimId = victimId.value
            event.energy = energy
            event.x = x
            event.y = y
            event.rammed = isRammed
        }
        return event
    }

    private fun map(botHitWallEvent: dev.robocode.tankroyale.server.event.BotHitWallEvent): BotHitWallEvent {
        val event = BotHitWallEvent()
        event.`$type` = Message.`$type`.BOT_HIT_WALL_EVENT
        botHitWallEvent.apply {
            event.turnNumber = turnNumber
            event.victimId = victimId.value
        }
        return event
    }

    private fun map(bulletFiredEvent: dev.robocode.tankroyale.server.event.BulletFiredEvent): BulletFiredEvent {
        val event = BulletFiredEvent()
        event.`$type` = Message.`$type`.BULLET_FIRED_EVENT
        bulletFiredEvent.apply {
            event.turnNumber = turnNumber
            event.bullet = BulletToBulletStateMapper.map(bullet)
        }
        return event
    }

    private fun map(bulletHitBotEvent: dev.robocode.tankroyale.server.event.BulletHitBotEvent): BulletHitBotEvent {
        val event = BulletHitBotEvent()
        event.`$type` = Message.`$type`.BULLET_HIT_BOT_EVENT
        bulletHitBotEvent.apply {
            event.turnNumber = turnNumber
            event.bullet = BulletToBulletStateMapper.map(bullet)
            event.victimId = victimId.value
            event.damage = damage
            event.energy = energy
        }
        return event
    }

    private fun map(bulletHitBulletEvent: dev.robocode.tankroyale.server.event.BulletHitBulletEvent): BulletHitBulletEvent {
        val event = BulletHitBulletEvent()
        event.`$type` = Message.`$type`.BULLET_HIT_BULLET_EVENT
        bulletHitBulletEvent.apply {
            event.turnNumber = turnNumber
            event.bullet = BulletToBulletStateMapper.map(bullet)
            event.hitBullet = BulletToBulletStateMapper.map(hitBullet)
        }
        return event
    }

    private fun map(bulletHitWallEvent: dev.robocode.tankroyale.server.event.BulletHitWallEvent): BulletHitWallEvent {
        val event = BulletHitWallEvent()
        event.`$type` = Message.`$type`.BULLET_HIT_WALL_EVENT
        bulletHitWallEvent.apply {
            event.turnNumber = turnNumber
            event.bullet = BulletToBulletStateMapper.map(bullet)
        }
        return event
    }

    private fun map(scannedBotEvent: dev.robocode.tankroyale.server.event.ScannedBotEvent): ScannedBotEvent {
        val event = ScannedBotEvent()
        event.`$type` = Message.`$type`.SCANNED_BOT_EVENT
        scannedBotEvent.apply {
            event.turnNumber = turnNumber
            event.scannedByBotId = scannedByBotId.value
            event.scannedBotId = scannedBotId.value
            event.energy = energy
            event.x = x
            event.y = y
            event.direction = direction
            event.speed = speed
        }
        return event
    }

    private fun map(skippedTurnEvent: dev.robocode.tankroyale.server.event.SkippedTurnEvent): SkippedTurnEvent {
        val event = SkippedTurnEvent()
        event.`$type` = Message.`$type`.SKIPPED_TURN_EVENT
        event.turnNumber = skippedTurnEvent.turnNumber
        return event
    }
}